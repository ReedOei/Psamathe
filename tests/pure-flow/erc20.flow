transformer newAccount(a : any nat) -> res : ! map any nat => any nat {
    [ ! record(key) { key : any nat, value : any nat } ;
        record(key) {
            key : any nat |-> a,
            value : any nat |-> 0
        }
    ] --> res
}

;

[ any nat ; 0,1,2 ] --> newAccount() --> var accounts : map any nat => any nat

4 --> accounts[0].value
5 --> accounts[1].value

accounts[0].value --> accounts[1].value
3 --> accounts[0].value

