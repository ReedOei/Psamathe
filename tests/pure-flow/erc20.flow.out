<generatedTop>
  <common>
    <typeEnv>
      .Map
    </typeEnv>
    <types>
      .Map
    </types>
  </common>
  <k>
    .
  </k>
  <functions>
    newAccount |-> transformer ( res : ! table ( key , .Vars ) ! record ( key , .Vars ) { key : any nat , value : any nat , .VarDefs } , a : any nat , .VarDefs , [ ! record ( key , .Vars ) { key : any nat , value : any nat , .VarDefs } ; record ( key , .Vars ) { key : any nat |-> a , value : any nat |-> 0 , .RecordMembers } , .Vars ] --> res  .Stmts )
  </functions>
  <lookup>
    accounts |-> selected ( loc ( 0 ) , loc ( 0 ) )
  </lookup>
  <storage>
    loc ( 0 ) |-> resource ( table ( key , .Vars ) ! record ( key , .Vars ) { key : any nat , value : any nat , .VarDefs } , ListItem ( loc ( 9 ) )
    ListItem ( loc ( 15 ) )
    ListItem ( loc ( 21 ) ) )
    loc ( 1 ) |-> resource ( nat , ListItem ( 0 ) )
    loc ( 2 ) |-> resource ( nat , ListItem ( 0 ) )
    loc ( 3 ) |-> resource ( nat , ListItem ( 0 ) )
    loc ( 4 ) |-> resource ( table ( key , .Vars ) ! record ( key , .Vars ) { key : any nat , value : any nat , .VarDefs } , .List )
    loc ( 5 ) |-> resource ( record ( key , .Vars ) { key : any nat , value : any nat , .VarDefs } , .List )
    loc ( 6 ) |-> resource ( nat , ListItem ( 0 ) )
    loc ( 7 ) |-> resource ( nat , ListItem ( 0 ) )
    loc ( 8 ) |-> resource ( nat , ListItem ( 3 ) )
    loc ( 9 ) |-> resource ( record ( key , .Vars ) { key : any nat , value : any nat , .VarDefs } , ListItem ( key |-> loc ( 6 )
    value |-> loc ( 8 ) ) )
    loc ( 10 ) |-> resource ( table ( key , .Vars ) ! record ( key , .Vars ) { key : any nat , value : any nat , .VarDefs } , .List )
    loc ( 11 ) |-> resource ( record ( key , .Vars ) { key : any nat , value : any nat , .VarDefs } , .List )
    loc ( 12 ) |-> resource ( nat , ListItem ( 1 ) )
    loc ( 13 ) |-> resource ( nat , ListItem ( 0 ) )
    loc ( 14 ) |-> resource ( nat , ListItem ( 9 ) )
    loc ( 15 ) |-> resource ( record ( key , .Vars ) { key : any nat , value : any nat , .VarDefs } , ListItem ( key |-> loc ( 12 )
    value |-> loc ( 14 ) ) )
    loc ( 16 ) |-> resource ( table ( key , .Vars ) ! record ( key , .Vars ) { key : any nat , value : any nat , .VarDefs } , .List )
    loc ( 17 ) |-> resource ( record ( key , .Vars ) { key : any nat , value : any nat , .VarDefs } , .List )
    loc ( 18 ) |-> resource ( nat , ListItem ( 2 ) )
    loc ( 19 ) |-> resource ( nat , ListItem ( 0 ) )
    loc ( 20 ) |-> resource ( nat , ListItem ( 0 ) )
    loc ( 21 ) |-> resource ( record ( key , .Vars ) { key : any nat , value : any nat , .VarDefs } , ListItem ( key |-> loc ( 18 )
    value |-> loc ( 20 ) ) )
    loc ( 22 ) |-> resource ( nat , ListItem ( 0 ) )
    loc ( 23 ) |-> resource ( nat , ListItem ( 0 ) )
    loc ( 24 ) |-> resource ( nat , ListItem ( 1 ) )
    loc ( 25 ) |-> resource ( nat , ListItem ( 0 ) )
    loc ( 26 ) |-> resource ( nat , ListItem ( 1 ) )
    loc ( 27 ) |-> resource ( nat , ListItem ( 0 ) )
    loc ( 28 ) |-> resource ( nat , ListItem ( 0 ) )
    loc ( 29 ) |-> resource ( nat , ListItem ( 0 ) )
  </storage>
  <catchStack>
    .List
  </catchStack>
</generatedTop>
